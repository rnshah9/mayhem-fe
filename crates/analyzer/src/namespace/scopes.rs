#![allow(unstable_name_collisions)] // expect_none, which ain't gonna be stabilized

use crate::context::{
    AnalyzerContext, CallType, Constant, ExpressionAttributes, FunctionBody, NamedThing,
};
use crate::errors::{AlreadyDefined, IncompleteItem, TypeError};
use crate::namespace::items::{EventId, FunctionId, ModuleId};
use crate::namespace::items::{Item, TypeDef};
use crate::namespace::types::{Type, TypeId};
use crate::AnalyzerDb;
use fe_common::diagnostics::Diagnostic;
use fe_common::Span;
use fe_parser::{ast, node::NodeId, Label};
use fe_parser::{ast::Expr, node::Node};
use indexmap::IndexMap;
use std::cell::RefCell;
use std::collections::BTreeMap;

pub struct ItemScope<'a> {
    db: &'a dyn AnalyzerDb,
    module: ModuleId,
    expressions: RefCell<IndexMap<NodeId, ExpressionAttributes>>,
    pub diagnostics: RefCell<Vec<Diagnostic>>,
}
impl<'a> ItemScope<'a> {
    pub fn new(db: &'a dyn AnalyzerDb, module: ModuleId) -> Self {
        Self {
            db,
            module,
            expressions: RefCell::new(IndexMap::default()),
            diagnostics: RefCell::new(vec![]),
        }
    }
}

impl<'a> AnalyzerContext for ItemScope<'a> {
    fn db(&self) -> &dyn AnalyzerDb {
        self.db
    }

    fn add_expression(&self, node: &Node<ast::Expr>, attributes: ExpressionAttributes) {
        self.expressions
            .borrow_mut()
            .insert(node.id, attributes)
            .expect_none("expression attributes already exist");
    }

    fn update_expression(&self, node: &Node<ast::Expr>, attributes: ExpressionAttributes) {
        self.expressions
            .borrow_mut()
            .insert(node.id, attributes)
            .expect("expression attributes do not exist");
    }

    fn expr_typ(&self, expr: &Node<Expr>) -> Type {
        self.expressions
            .borrow()
            .get(&expr.id)
            .unwrap()
            .typ
            .typ(self.db())
    }

    fn add_constant(
        &self,
        _name: &Node<ast::SmolStr>,
        _expr: &Node<ast::Expr>,
        _value: crate::context::Constant,
    ) {
        // We use salsa query to get constant. So no need to add constant
        // explicitly.
    }

    fn constant_value_by_name(
        &self,
        name: &ast::SmolStr,
        _span: Span,
    ) -> Result<Option<Constant>, IncompleteItem> {
        if let Some(constant) = self.module.resolve_constant(self.db, name)? {
            // It's ok to ignore an error.
            // Diagnostics are already emitted when an error occurs.
            Ok(constant.constant_value(self.db).ok())
        } else {
            Ok(None)
        }
    }

    fn parent(&self) -> Item {
        Item::Module(self.module)
    }

    fn module(&self) -> ModuleId {
        self.module
    }

    fn parent_function(&self) -> FunctionId {
        panic!("ItemContext has no parent function")
    }

    fn add_call(&self, _node: &Node<ast::Expr>, _call_type: CallType) {
        unreachable!("Can't call function outside of function")
    }

    fn is_in_function(&self) -> bool {
        false
    }

    fn inherits_type(&self, _typ: BlockScopeType) -> bool {
        false
    }

    fn resolve_name(&self, name: &str, span: Span) -> Result<Option<NamedThing>, IncompleteItem> {
        let item = self.module.resolve_name(self.db, name)?;

        if let Some(item) = item {
            check_item_visibility(self, item, span);
        }

        Ok(item.map(NamedThing::Item))
    }

    fn resolve_path(&self, path: &ast::Path, span: Span) -> Option<NamedThing> {
        let item = self.module.resolve_path_internal(self.db(), path);

        for diagnostic in item.diagnostics.iter() {
            self.add_diagnostic(diagnostic.clone())
        }

        if let Some(item) = item.value {
            check_item_visibility(self, item, span);
        }

        item.value.map(NamedThing::Item)
    }

    fn add_diagnostic(&self, diag: Diagnostic) {
        self.diagnostics.borrow_mut().push(diag)
    }

    /// Gets `std::context::Context` if it exists
    fn get_context_type(&self) -> Option<TypeId> {
        if let Ok(Some(NamedThing::Item(Item::Type(TypeDef::Struct(id))))) =
            // `Context` is guaranteed to be public, so we can use `Span::dummy()` .
            self.resolve_name("Context", Span::dummy())
        {
            // we just assume that there is only one `Context` defined in `std`
            if id.module(self.db()).ingot(self.db()).name(self.db()) == "std" {
                return Some(self.db().intern_type(Type::Struct(id)));
            }
        }

        None
    }
}

pub struct FunctionScope<'a> {
    pub db: &'a dyn AnalyzerDb,
    pub function: FunctionId,
    pub body: RefCell<FunctionBody>,
    pub diagnostics: RefCell<Vec<Diagnostic>>,
}

impl<'a> FunctionScope<'a> {
    pub fn new(db: &'a dyn AnalyzerDb, function: FunctionId) -> Self {
        Self {
            db,
            function,
            body: RefCell::new(FunctionBody::default()),
            diagnostics: RefCell::new(vec![]),
        }
    }

    pub fn function_return_type(&self) -> Result<TypeId, TypeError> {
        self.function.signature(self.db).return_type.clone()
    }

    /// Attribute contextual information to an emit statement node.
    ///
    /// # Panics
    ///
    /// Panics if an entry already exists for the node id.
    pub fn add_emit(&self, node: &Node<ast::FuncStmt>, event: EventId) {
        self.add_node(node);
        self.body
            .borrow_mut()
            .emits
            .insert(node.id, event)
            .expect_none("emit statement attributes already exist");
    }

    pub fn map_variable_type<T>(&self, node: &Node<T>, typ: TypeId) {
        self.add_node(node);
        self.body
            .borrow_mut()
            .var_types
            .insert(node.id, typ)
            .expect_none("variable has already registered")
    }

    fn add_node<T>(&self, node: &Node<T>) {
        self.body.borrow_mut().spans.insert(node.id, node.span);
    }
}

impl<'a> AnalyzerContext for FunctionScope<'a> {
    fn db(&self) -> &dyn AnalyzerDb {
        self.db
    }

    fn add_diagnostic(&self, diag: Diagnostic) {
        self.diagnostics.borrow_mut().push(diag)
    }

    fn add_expression(&self, node: &Node<ast::Expr>, attributes: ExpressionAttributes) {
        self.add_node(node);
        self.body
            .borrow_mut()
            .expressions
            .insert(node.id, attributes)
            .expect_none("expression attributes already exist");
    }

    fn update_expression(&self, node: &Node<ast::Expr>, attributes: ExpressionAttributes) {
        self.body
            .borrow_mut()
            .expressions
            .insert(node.id, attributes)
            .expect("expression attributes do not exist");
    }

    fn expr_typ(&self, expr: &Node<Expr>) -> Type {
        self.body
            .borrow()
            .expressions
            .get(&expr.id)
            .unwrap()
            .typ
            .typ(self.db())
    }

    fn add_constant(&self, _name: &Node<ast::SmolStr>, expr: &Node<ast::Expr>, value: Constant) {
        self.body
            .borrow_mut()
            .expressions
            .get_mut(&expr.id)
            .expect("expression attributes must exist before adding constant value")
            .const_value = Some(value);
    }

    fn constant_value_by_name(
        &self,
        _name: &ast::SmolStr,
        _span: Span,
    ) -> Result<Option<Constant>, IncompleteItem> {
        Ok(None)
    }

    fn parent(&self) -> Item {
        self.function.parent(self.db())
    }

    fn module(&self) -> ModuleId {
        self.function.module(self.db())
    }

    fn parent_function(&self) -> FunctionId {
        self.function
    }

    fn add_call(&self, node: &Node<ast::Expr>, call_type: CallType) {
        // TODO: should probably take the Expr::Call node, rather than the function node

        self.add_node(node);
        self.body
            .borrow_mut()
            .calls
            .insert(node.id, call_type)
            .expect_none("call attributes already exist");
    }

    fn is_in_function(&self) -> bool {
        true
    }

    fn inherits_type(&self, _typ: BlockScopeType) -> bool {
        false
    }

    fn resolve_name(&self, name: &str, span: Span) -> Result<Option<NamedThing>, IncompleteItem> {
        let sig = self.function.signature(self.db);

        if name == "self" {
            return Ok(Some(NamedThing::SelfValue {
                decl: sig.self_decl,
                parent: self.function.sig(self.db).self_item(self.db),
                span: self.function.self_span(self.db),
            }));
        }

        // Getting param names and spans should be simpler
        let param = sig.params.iter().find_map(|param| {
            (param.name == name).then(|| {
                let span = self
                    .function
                    .data(self.db)
                    .ast
                    .kind
                    .sig
                    .kind
                    .args
                    .iter()
                    .find_map(|param| (param.name() == name).then(|| param.name_span()))
                    .expect("found param type but not span");

                NamedThing::Variable {
                    name: name.to_string(),
                    typ: param.typ.clone(),
                    is_const: false,
                    span,
                }
            })
        });

        if let Some(param) = param {
            Ok(Some(param))
        } else {
            let item =
                if let Item::Type(TypeDef::Contract(contract)) = self.function.parent(self.db) {
                    contract.resolve_name(self.db, name)
                } else {
                    self.function.module(self.db).resolve_name(self.db, name)
                }?;

            if let Some(item) = item {
                check_item_visibility(self, item, span);
            }

            Ok(item.map(NamedThing::Item))
        }
    }

    fn resolve_path(&self, path: &ast::Path, span: Span) -> Option<NamedThing> {
        let item = self
            .function
            .module(self.db())
            .resolve_path_internal(self.db(), path);

        for diagnostic in item.diagnostics.iter() {
            self.add_diagnostic(diagnostic.clone())
        }

        if let Some(item) = item.value {
            check_item_visibility(self, item, span);
        }

        item.value.map(NamedThing::Item)
    }

    fn get_context_type(&self) -> Option<TypeId> {
        if let Ok(Some(NamedThing::Item(Item::Type(TypeDef::Struct(id))))) =
            self.resolve_name("Context", Span::dummy())
        {
            if id.module(self.db()).ingot(self.db()).name(self.db()) == "std" {
                return Some(self.db().intern_type(Type::Struct(id)));
            }
        }

        None
    }
}

pub struct BlockScope<'a, 'b> {
    pub root: &'a FunctionScope<'b>,
    pub parent: Option<&'a BlockScope<'a, 'b>>,
    /// Maps Name -> (Type, is_const, span)
    pub variable_defs: BTreeMap<String, (TypeId, bool, Span)>,
    pub constant_defs: RefCell<BTreeMap<String, Constant>>,
    pub typ: BlockScopeType,
}

#[derive(Clone, Debug, PartialEq)]
pub enum BlockScopeType {
    Function,
    IfElse,
    Loop,
    Unsafe,
}

impl AnalyzerContext for BlockScope<'_, '_> {
    fn db(&self) -> &dyn AnalyzerDb {
        self.root.db
    }

    fn resolve_name(&self, name: &str, span: Span) -> Result<Option<NamedThing>, IncompleteItem> {
        if let Some(var) =
            self.variable_defs
                .get(name)
                .map(|(typ, is_const, span)| NamedThing::Variable {
                    name: name.to_string(),
                    typ: Ok(*typ),
                    is_const: *is_const,
                    span: *span,
                })
        {
            Ok(Some(var))
        } else if let Some(parent) = self.parent {
            parent.resolve_name(name, span)
        } else {
            self.root.resolve_name(name, span)
        }
    }

    fn add_expression(&self, node: &Node<ast::Expr>, attributes: ExpressionAttributes) {
        self.root.add_expression(node, attributes)
    }

    fn update_expression(&self, node: &Node<ast::Expr>, attributes: ExpressionAttributes) {
        self.root.update_expression(node, attributes)
    }

    fn expr_typ(&self, expr: &Node<Expr>) -> Type {
        self.root.expr_typ(expr)
    }

    fn add_constant(&self, name: &Node<ast::SmolStr>, expr: &Node<ast::Expr>, value: Constant) {
        self.constant_defs
            .borrow_mut()
            .insert(name.kind.clone().to_string(), value.clone())
            .expect_none("expression attributes already exist");

        self.root.add_constant(name, expr, value)
    }

    fn constant_value_by_name(
        &self,
        name: &ast::SmolStr,
        span: Span,
    ) -> Result<Option<Constant>, IncompleteItem> {
        if let Some(constant) = self.constant_defs.borrow().get(name.as_str()) {
            Ok(Some(constant.clone()))
        } else if let Some(parent) = self.parent {
            parent.constant_value_by_name(name, span)
        } else {
            match self.resolve_name(name, span)? {
                Some(NamedThing::Item(Item::Constant(constant))) => {
                    Ok(constant.constant_value(self.db()).ok())
                }
                _ => Ok(None),
            }
        }
    }

    fn parent(&self) -> Item {
        Item::Function(self.root.function)
    }

    fn module(&self) -> ModuleId {
        self.root.module()
    }

    fn parent_function(&self) -> FunctionId {
        self.root.function
    }

    fn add_call(&self, node: &Node<ast::Expr>, call_type: CallType) {
        self.root.add_call(node, call_type)
    }

    fn is_in_function(&self) -> bool {
        true
    }

    fn inherits_type(&self, typ: BlockScopeType) -> bool {
        self.typ == typ || self.parent.map_or(false, |scope| scope.inherits_type(typ))
    }

    fn resolve_path(&self, path: &ast::Path, span: Span) -> Option<NamedThing> {
        self.root.resolve_path(path, span)
    }

    fn add_diagnostic(&self, diag: Diagnostic) {
        self.root.diagnostics.borrow_mut().push(diag)
    }

    fn get_context_type(&self) -> Option<TypeId> {
        self.root.get_context_type()
    }
}

impl<'a, 'b> BlockScope<'a, 'b> {
    pub fn new(root: &'a FunctionScope<'b>, typ: BlockScopeType) -> Self {
        BlockScope {
            root,
            parent: None,
            variable_defs: BTreeMap::new(),
            constant_defs: RefCell::new(BTreeMap::new()),
            typ,
        }
    }

    pub fn new_child(&'a self, typ: BlockScopeType) -> Self {
        BlockScope {
            root: self.root,
            parent: Some(self),
            variable_defs: BTreeMap::new(),
            constant_defs: RefCell::new(BTreeMap::new()),
            typ,
        }
    }

    /// Add a variable to the block scope.
    pub fn add_var(
        &mut self,
        name: &str,
        typ: TypeId,
        is_const: bool,
        span: Span,
    ) -> Result<(), AlreadyDefined> {
        match self.resolve_name(name, span) {
            Ok(Some(NamedThing::SelfValue { .. })) => {
                self.error(
                    "`self` can't be used as a variable name",
                    span,
                    "expected a name, found keyword `self`",
                );
                Err(AlreadyDefined)
            }

            Ok(Some(named_item)) => {
                if named_item.is_builtin() {
                    self.error(
                        &format!(
                            "variable name conflicts with built-in {}",
                            named_item.item_kind_display_name(),
                        ),
                        span,
                        &format!(
                            "`{}` is a built-in {}",
                            name,
                            named_item.item_kind_display_name()
                        ),
                    );
                    return Err(AlreadyDefined);
                } else {
                    // It's (currently) an error to shadow a variable in a nested scope
                    self.duplicate_name_error(
                        &format!("duplicate definition of variable `{}`", name),
                        name,
                        named_item
                            .name_span(self.db())
                            .expect("missing name_span of non-builtin"),
                        span,
                    );
                }
                Err(AlreadyDefined)
            }
            _ => {
                self.variable_defs
                    .insert(name.to_string(), (typ, is_const, span));
                Ok(())
            }
        }
    }
}

/// temporary helper until `BTreeMap::try_insert` is stabilized
trait OptionExt {
    fn expect_none(self, msg: &str);
}

impl<T> OptionExt for Option<T> {
    fn expect_none(self, msg: &str) {
        if self.is_some() {
            panic!("{}", msg)
        }
    }
}

/// Check an item visibility and sink diagnostics if an item is invisible from
/// the scope.
fn check_item_visibility(context: &dyn AnalyzerContext, item: Item, span: Span) {
    let item_module = item
        .module(context.db())
        .unwrap_or_else(|| context.module());
    if !item.is_public(context.db()) && item_module != context.module() {
        let module_name = item_module.name(context.db());
        let item_name = item.name(context.db());
        let item_span = item.name_span(context.db()).unwrap_or(span);
        let item_kind_name = item.item_kind_display_name();
        context.fancy_error(
            &format!("the {} `{}` is private", item_kind_name, item_name,),
            vec![
                Label::primary(span, format!("this {} is not `pub`", item_kind_name)),
                Label::secondary(item_span, format!("`{}` is defined here", item_name)),
            ],
            vec![
                format!("`{}` can only be used within `{}`", item_name, module_name),
                format!(
                    "Hint: use `pub` to make `{}` visible from outside of `{}`",
                    item_name, module_name,
                ),
            ],
        );
    }
}
