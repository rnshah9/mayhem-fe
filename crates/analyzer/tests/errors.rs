//! Tests for contracts that should cause compile errors

use fe_analyzer::namespace::items::{IngotId, IngotMode, ModuleId};
use fe_analyzer::TestDb;
use fe_common::diagnostics::diagnostics_string;
use fe_common::files::FileKind;
use indexmap::indexmap;
use insta::assert_snapshot;
use wasm_bindgen_test::wasm_bindgen_test;

fn error_string(path: &str, src: &str) -> String {
    let mut db = TestDb::default();
    let module = ModuleId::new_standalone(&mut db, path, src);

    let diags = module.diagnostics(&db);
    if diags.is_empty() {
        panic!("expected analysis to fail with an error")
    }
    diagnostics_string(&db, &diags)
}

fn error_string_ingot(path: &str) -> String {
    let mut db = TestDb::default();
    let std = IngotId::std_lib(&mut db);
    let ingot = IngotId::from_files(
        &mut db,
        "test_ingot",
        IngotMode::Main,
        FileKind::Local,
        &test_files::fixture_dir_files(path),
        indexmap! { "std".into() => std },
    );

    let diags = ingot.diagnostics(&db);
    if diags.is_empty() {
        panic!("expected analysis to fail with an error");
    }
    diagnostics_string(&db, &diags)
}

macro_rules! test_ingot {
    ($name:ident) => {
        #[test]
        #[wasm_bindgen_test]
        fn $name() {
            let path = concat!("compile_errors/", stringify!($name), "/src");

            if cfg!(target_arch = "wasm32") {
                fe_common::assert_snapshot_wasm!(
                    concat!("snapshots/errors__", stringify!($name), ".snap"),
                    error_string_ingot(&path)
                );
            } else {
                assert_snapshot!(error_string_ingot(&path));
            }
        }
    };
}

macro_rules! test_file {
    ($name:ident) => {
        #[test]
        #[wasm_bindgen_test]
        fn $name() {
            let path = concat!("compile_errors/", stringify!($name), ".fe");
            if cfg!(target_arch = "wasm32") {
                fe_common::assert_snapshot_wasm!(
                    concat!("snapshots/errors__", stringify!($name), ".snap"),
                    error_string(&path, test_files::fixture(path))
                );
            } else {
                assert_snapshot!(error_string(&path, test_files::fixture(path)));
            }
        }
    };
}

macro_rules! test_stmt {
    ($name:ident, $stmt:expr) => {
        #[test]
        #[wasm_bindgen_test]
        fn $name() {
            let src = format!(
                "contract C {{\n pub fn f(self) {{\n  {}\n }}\n}}",
                $stmt.replace('\n', "\n  ")
            );
            if cfg!(target_arch = "wasm32") {
                fe_common::assert_snapshot_wasm!(
                    concat!("snapshots/errors__", stringify!($name), ".snap"),
                    error_string("[snippet]", &src)
                );
            } else {
                assert_snapshot!(error_string("[snippet]", &src));
            }
        }
    };
}

test_stmt! { array_mixed_types, "let x: Array<u16, 3> = [1, address(0), \"hi\"]" }
test_stmt! { array_size_mismatch, "let x: Array<u8, 3> = []\nlet y: Array<u8, 3> = [1, 2]" }
test_stmt! { array_constructor_call, "u8[3]([1, 2, 3])" }
test_stmt! { assert_reason_not_string, "assert true, 1" }
test_stmt! { assign_int, "5 = 6" }
test_stmt! { assign_call, "self.f() = 10" }
test_stmt! { assign_type_mismatch, "let x: u256 = 10\nx = address(0)" }
test_stmt! { aug_assign_non_numeric, "let a: u256 = 1\nlet b: bool = true\na += b" }
test_stmt! { binary_op_add_uints, "let a: u256 = 1\nlet b: u8 = 2\na + b" }
test_stmt! { binary_op_lshift_bool, "let a: bool = true\nlet b: i256\na << b" }
test_stmt! { binary_op_lshift_with_int, "let a: u256 = 1\nlet b: i256 = 2\na << b" }
test_stmt! { binary_op_pow_int, "let a: u256 = 1\nlet b: i256 = 2\na ** b" }
test_stmt! { binary_op_boolean_mismatch1, "10 and true" }
test_stmt! { binary_op_boolean_mismatch2, "false or 1" }
test_stmt! { binary_op_boolean_mismatch3, "1 or 2" }
test_stmt! { bool_constructor, "bool(true)" }
test_stmt! { bool_cast, "bool(0)" }
test_stmt! { break_without_loop, "break" }
test_stmt! { break_without_loop_2, "if true { break }" }
test_stmt! { call_undefined_function_on_contract, "self.doesnt_exist()" }
test_stmt! { call_address_with_wrong_type, "address(true)" }
test_stmt! { call_address_with_label, "address(val: 0)" }
test_stmt! { call_keccak_without_parameter, "keccak256()" }
test_stmt! { call_keccak_with_wrong_type, "keccak256(true)" }
test_stmt! { call_keccak_with_2_args, "keccak256(1, 2)" }
test_stmt! { call_keccak_with_generic_args, "keccak256<10>(1)" }
test_stmt! { cast_address_to_u64, "u64(address(0))" }

test_stmt! { call_balance_of_without_parameter, "unsafe { std::evm::balance_of() }" }
test_stmt! { call_balance_of_with_wrong_type, "unsafe { std::evm::balance_of(true) }" }
test_stmt! { call_balance_of_with_2_args, "unsafe { std::evm::balance_of(address(0), 2) }" }
test_stmt! { call_balance_with_arg, "unsafe { std::evm::balance(address(0)) }" }
test_stmt! { clone_arg_count, "let x: Array<u256, 2> = [5, 6]\nlet y: Array<u256, 2> = x.clone(y)" }
test_stmt! { continue_without_loop, "continue" }
test_stmt! { continue_without_loop_2, "if true { continue }" }
test_stmt! { emit_undefined_event, "emit MyEvent()" }
test_stmt! { emit_type_name, "emit u8()" }
test_stmt! { emit_variable, "let x: u8 = 10\nemit x()" }
test_stmt! { int_type_generic_arg_list, "let x: u256<>" }
test_stmt! { int_type_generic_arg, "let x: u256<10>" }
test_stmt! { int_type_constructor_generic_arg_list, "u256<>(10)" }
test_stmt! { int_type_constructor_generic_arg, "u256<1>(10)" }
test_stmt! { map_three_type_args, "let x: Map<u8, u8, u8>" }
test_stmt! { map_int_type_arg, "let x: Map<address, 100>" }
test_stmt! { map_int_type_args, "let x: Map<10, 20>" }
test_stmt! { map_no_type_args, "let x: Map<>" }
test_stmt! { map_no_type_arg_list, "let x: Map" }
test_stmt! { map_one_type_arg, "let x: Map<y>" }
test_stmt! { map_map_key_type, "let x: Map<Map<u8, u8>, address>" }
test_stmt! { map_constructor, "Map<u8, u8>()" }
test_stmt! { non_bool_and, "let x: bool = true\nlet y: u256 = 1\nx = x and y" }
test_stmt! { non_bool_or, "let x: bool = true\nlet y: u256 = 1\nx = x or y" }
test_stmt! { overflow_i128_neg, "i128(-170141183460469231731687303715884105729)" }
test_stmt! { overflow_i128_pos, "i128(170141183460469231731687303715884105728)" }
test_stmt! { overflow_i16_neg, "i16(-32769)" }
test_stmt! { overflow_i16_pos, "i16(32768)" }
test_stmt! { overflow_i256_neg, "i256(-57896044618658097711785492504343953926634992332820282019728792003956564819969)" }
test_stmt! { overflow_i256_pos, "i256(57896044618658097711785492504343953926634992332820282019728792003956564819968)" }
test_stmt! { overflow_i32_neg, "i32(-2147483649)" }
test_stmt! { overflow_i32_pos, "i32(2147483648)" }
test_stmt! { overflow_i64_neg, "i64(-9223372036854775809)" }
test_stmt! { overflow_i64_pos, "i64(9223372036854775808)" }
test_stmt! { overflow_i8_neg, "let x: i8 = -129" }
test_stmt! { overflow_i8_pos, "let x: i8 = 128" }
test_stmt! { overflow_literal_too_big, "115792089237316195423570985008687907853269984665640564039457584007913129639936" }
test_stmt! { overflow_literal_too_small, "-115792089237316195423570985008687907853269984665640564039457584007913129639936" }
test_stmt! { overflow_u128_neg, "let x: u128 = -1" }
test_stmt! { overflow_u128_pos, "u128(340282366920938463463374607431768211456)" }
test_stmt! { overflow_u16_neg, "u16(-1)" }
test_stmt! { overflow_u16_pos, "u16(65536)" }
test_stmt! { overflow_u256_neg, "u256(-1)" }
test_stmt! { overflow_u256_pos, "u256(115792089237316195423570985008687907853269984665640564039457584007913129639936)" }
test_stmt! { overflow_u32_neg, "u32(-1)" }
test_stmt! { overflow_u32_pos, "u32(4294967296)" }
test_stmt! { overflow_u64_neg, "u64(-1)" }
test_stmt! { overflow_u64_pos, "u64(18446744073709551616)" }
test_stmt! { overflow_u8_neg, "u8(-1)" }
test_stmt! { overflow_u8_pos, "u8(256)" }
test_stmt! { overflow_u8_assignment, "let x: u8 = 260" }
test_stmt! { pow_with_signed_exponent, "let base: i128\nlet xp: i128\nbase ** exp" }
// Exponent can be unsigned but needs to be same size or smaller
test_stmt! { pow_with_wrong_capacity, "let base: i128\nlet exp: u256\nbase ** exp" }
test_stmt! { shadow_builtin_type_with_var, "let u8: u8 = 10" }
test_stmt! { shadow_builtin_fn_with_var, "let keccak256: u8 = 10" }
test_file! { shadow_builtin_type }
test_file! { shadow_builtin_function }
test_file! { self_misuse }
test_stmt! { string_capacity_mismatch, "String<3>(\"too long\")" }
test_stmt! { string_non_int_type_arg, "let x: String<u8>" }
test_stmt! { string_no_type_arg_list, "let x: String" }
test_stmt! { string_no_type_args, "let x: String<>" }
test_stmt! { string_two_int_type_args, "let x: String<1, 2>" }
test_stmt! { string_two_type_args, "let x: String<1, u8>" }
test_stmt! { string_constructor_non_int_type_arg, "String<foo>()" } // issue #532
test_stmt! { string_constructor_no_type_arg_list, "String()" }
test_stmt! { string_constructor_no_type_args, "String<>()" }
test_stmt! { string_constructor_two_int_type_args, "String<1, 2>()" }
test_stmt! { string_constructor_two_type_args, "String<1, u8>()" }
test_stmt! { ternary_type_mismatch, "10 if 100 else true" }
test_stmt! { change_sign_and_size_in_cast, "let x: i8\nlet y: u16 = u16(x)" }
test_stmt! { change_sign_and_type_in_cast, "let x: bool\nlet y: u16 = u16(x)" }
test_stmt! { type_constructor_arg_count, "let x: u8 = u8(1, 10)" }
test_stmt! { unary_minus_on_bool, "let x: bool = true\n-x" }
test_stmt! { unary_not_on_int, "let x: u256 = 10\nnot x" }
test_stmt! { unary_always_mismatch_type_case_1, "let x:u256 = 10\nlet y:u256 = -x"}
test_stmt! { unary_always_mismatch_type_case_2, "let x:i32 = -10\nlet y:u256 = -x"}
test_stmt! { undefined_generic_type, "let x: foobar<u256> = 10" }
test_stmt! { undefined_name, "let x: u16 = y\nlet z: u16 = y\nlet v: u16 = _42" }
test_stmt! { undefined_type, "let x: foobar = 10" }
test_stmt! { unexpected_return, "return 1" }
test_stmt! { unit_type_constructor, "()()" }
test_stmt! { revert_reason_not_struct, "revert 1" }
test_stmt! { invalid_ascii, "String<2>(\"ä\")" }
test_stmt! { invert_non_numeric, "~true" }

test_file! { ambiguous_traits }
test_file! { ambiguous_traits2 }
test_ingot! { trait_not_in_scope }
test_file! { bad_string }
test_file! { bad_tuple_attr1 }
test_file! { bad_tuple_attr2 }
test_file! { bad_tuple_attr3 }
test_file! { call_generic_function_with_unsatisfied_bound}
test_file! { call_builtin_object }
test_file! { call_create_with_wrong_type }
test_file! { call_create2_with_wrong_type }
test_file! { call_event_with_wrong_types }
test_file! { call_static_function_without_double_colon }
test_file! { call_undefined_function_on_external_contract }
test_file! { call_undefined_function_on_memory_struct }
test_file! { call_undefined_function_on_storage_struct }
test_file! { call_non_pub_fn_on_external_contract }
test_file! { cannot_move }
test_file! { cannot_move2 }
test_file! { circular_dependency_create }
test_file! { circular_dependency_create2 }
test_file! { circular_type_alias }
test_file! { const_assign }
test_file! { const_generics_param }
test_file! { const_local }
test_file! { duplicate_arg_in_contract_method }
test_file! { duplicate_contract_in_module }
test_file! { duplicate_event_in_contract }
test_file! { duplicate_field_in_contract }
test_file! { duplicate_field_in_struct }
test_file! { duplicate_method_in_contract }
test_file! { duplicate_struct_in_module }
test_file! { duplicate_typedef_in_module }
test_file! { duplicate_var_in_child_scope }
test_file! { duplicate_var_in_contract_method }
test_file! { duplicate_var_in_for_loop }
test_file! { duplicate_generic_params }
test_file! { emit_bad_args }
test_file! { external_call_type_error }
test_file! { external_call_wrong_number_of_params }
test_file! { contract_function_with_generic_params }
test_file! { indexed_event }
test_file! { invalid_compiler_version }
test_file! { invalid_block_field }
test_file! { invalid_chain_field }
test_file! { invalid_contract_field }
test_file! { invalid_generic_bound }
test_file! { invalid_impl_type }
test_file! { invalid_impl_location }
test_file! { invalid_msg_field }
test_file! { invalid_string_field }
test_file! { invalid_struct_field }
test_file! { invalid_tuple_field }
test_file! { invalid_tx_field }
test_file! { invalid_var_declaration_1 }
test_file! { invalid_var_declaration_2 }
test_file! { issue_451 }
test_file! { mislabeled_call_args }
test_file! { mislabeled_call_args_self }
test_file! { mislabeled_call_args_external_contract_call }
test_file! { mismatch_return_type }
test_file! { missing_return }
test_file! { missing_return_in_else }
test_file! { missing_return_after_if }
test_file! { module_const_unknown_type }
test_file! { module_const_non_base_type }
test_file! { module_const_call }
test_file! { needs_mem_copy }
test_file! { not_callable }
test_file! { not_in_scope }
test_file! { not_in_scope_2 }
test_file! { private_struct_field }
test_file! { return_addition_with_mixed_types }
test_file! { return_call_to_fn_with_param_type_mismatch }
test_file! { return_call_to_fn_without_return }
test_file! { return_from_init }
test_file! { return_lt_mixed_types }
test_file! { return_type_undefined }
test_file! { return_type_not_fixedsize }
test_file! { undefined_type_param }

test_file! { strict_boolean_if_else }
test_file! { struct_private_constructor }
test_file! { struct_call_bad_args }
test_file! { struct_call_without_kw_args }
test_file! { struct_recursive_cycles }
test_file! { trait_impl_mismatch }
test_file! { trait_fn_without_self }
test_file! { trait_fn_with_generic_params }
test_file! { traits_as_fields }
test_file! { trait_conflicting_impls }
test_file! { traits_with_wrong_bounds }
test_file! { non_pub_init }
test_file! { init_wrong_return_type }
test_file! { init_duplicate_def }
test_file! { init_call_on_self }
test_file! { init_call_on_external_contract }
test_file! { call_wrong_return_type }
test_file! { call_duplicate_def }
test_file! { call_call_on_self }
test_file! { call_call_on_external_contract }
test_file! { call_with_pub_fns }
test_file! { abi_encode_u256 }
test_file! { abi_encode_from_storage }
test_file! { assert_sto_msg_no_copy }
test_file! { for_loop_sto_iter_no_copy }
test_file! { revert_sto_error_no_copy }

test_file! { call_to_mut_fn_without_self }
test_file! { call_to_pure_fn_on_self }
test_file! { missing_self }
test_file! { self_not_first }
test_file! { self_in_standalone_fn }
test_file! { unsafe_misuse }
test_file! { unsafe_nesting }

test_ingot! { bad_ingot }
test_ingot! { mainless_ingot }
test_ingot! { bad_visibility }

test_file! { ctx_not_first }
test_file! { ctx_not_after_self }
test_file! { ctx_init }
test_file! { ctx_pure }
test_file! { ctx_undeclared }
test_file! { ctx_missing_internal_call }
test_file! { ctx_missing_create }
test_file! { ctx_missing_load }
test_file! { ctx_missing_event }
test_file! { ctx_builtins_param_incorrect_type }
test_file! { ctx_undefined_create }
test_file! { ctx_undefined_create2 }
test_file! { ctx_undefined_event }
test_file! { uninit_values }
test_file! { invalid_repeat_length }
