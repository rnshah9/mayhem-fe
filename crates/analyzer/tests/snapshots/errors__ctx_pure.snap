---
source: crates/analyzer/tests/errors.rs
expression: "error_string(&path, test_files::fixture(path))"

---
error: `ctx` cannot be passed into pure functions
  ┌─ compile_errors/ctx_pure.fe:3:12
  │
3 │ pub fn foo(ctx: Context) {}
  │            ^^^^^^^^^^^^ `ctx` can only be passed into contract functions


